### ISSUE TITLE
[Feature request] volume control

### ISSUE THREAD

**Issue by Anonymous941** [LumaTeam/Luma3DS#1913]
I have an old 2DS, and the volume slider is broken, and it appears I'm [not](https://gbatemp.net/threads/nintendo-2ds-volume-slider-broken-and-stuck-on-the-loudest-settings.382529/) [the](https://gbatemp.net/threads/2ds-volume-slider-repair.605898/) [only](https://gbatemp.net/threads/n2ds-xl-volume-slider-repair.531757/) [one](https://www.ifixit.com/Answers/View/691832/How+would+I+fix+a+volume+button).  My first thought was that this isn't software-controlled, but when using nds-bootstrap, I'm able to adjust it.  Also when I move the console too much, the volume goes back to 83% (where the slider is stuck) until I go into the menu and change it again.

It appears that volume control is in `retail/cardenginei/arm9_igm/source/inGameMenu.c`:

```
case OPTIONS_VOLUME:
{                                                                                                                                                                                                        
    u8 volume = (u8)(sharedAddr[6] >> 16);
    if(KEYS & KEY_LEFT && volume > 0) 
        volume--;
        else if (KEYS & KEY_RIGHT && volume < 31)
            volume++;

            sharedAddr[0] = volume;
            sharedAddr[4] = 0x554C4F56; // VOLU
            while(sharedAddr[4] == 0x554C4F56) {
                while (REG_VCOUNT != 191) swiDelay(100);
                while (REG_VCOUNT == 191) swiDelay(100);
        }
        break;
}
```

Can this be added to the Rosalina menu, possibly with the option to periodically re-change it?

**Comment by Anonymous941** [comment]
I found [an outdated fork](https://github.com/LumaTeam/Luma3DS/commit/dc636d82492d1e87eb51785fa7f2a98617e7ece9) ([commit 2](https://github.com/LumaTeam/Luma3DS/commit/a9b05f35a9c0c3f45abab2f0b57a39309461a8fe)) that has this feature, so it can be added fairly easily...

**Comment by ShelterPush** [comment]
I am uncertain if it warrants the creation of another feature request, but in addition to changing the volume level, could this feature include the ability to change the number of channels (mono - stereo - surround)? From what I understand of it, the reference commit only controls volume itself.

I frequently use a single earbud for games, but I would like to be able to switch them to stereo or surround when I take out the earbuds without having to shut down the game I'm playing.

**Comment by cooolgamer** [comment]
> I found [an outdated fork](https://github.com/LumaTeam/Luma3DS/commit/dc636d82492d1e87eb51785fa7f2a98617e7ece9) ([commit 2](https://github.com/LumaTeam/Luma3DS/commit/a9b05f35a9c0c3f45abab2f0b57a39309461a8fe)) that has this feature, so it can be added fairly easily...

I added that to my fork and it works fine on the latest luma version, you just have to change some things because just copy/paste it doesn't work. So yes, it is possible to have this on the latest luma version

**Comment by Anonymous941** [comment]
@cooolgamer Mind putting it in a PR?

**Comment by cooolgamer** [comment]
I did not made it, but I put it on luma v13, I will make a pull request with this when I can (not at home rn) and ofc crediting the original creators (Sono and Nutez)

**Comment by profi200** [comment]
Fwiw the MCU is not a good place to override volume. If the slider is wonky these regs will not work right. On CODEC however you have full software control via the shutter sound mechanism which is just overriding the slider temporarily.

**Comment by Anonymous941** [comment]
What would happen when the shutter sound triggered?  Would it revert back to the previous volume once it's done?

**Comment by profi200** [comment]
I have been experimenting with this a bit and got it to work. Shutter sound only happens via I2S2 which is not responsible for DSP sound. I2S1 volume override is not touched by codec module as far as i can tell so the camera app should do what it always does without affecting this setting.

I didn't figure out yet how the volume slider value maps to these volume control registers since it has a differnt range. -128 - 48. And it's in 0.5 dB steps (-128 is muted which leaves us with a range of -63.5 - +24 dB range).

**Comment by profi200** [comment]
Here is the proper software volume control code. It sticks even in sleep mode. It does not survive reboots so it must be set each boot. Camera shutter sound is unaffected and the shutter mechanism used in codec module does also not affect this volume override.

By trial and error i found out the maximum volume slider position maps to volume value -20 (-10 dB). I highly recommend not going above this value because i don't know if this will overload the headphone/speaker drivers.

```
s8 volumes[2] = {-64, -64}; // Volume in 0.5 dB steps. -128 (muted) to 48. Do not go above -20 (100%).
u8 tmp;
cdcChkInit();
CDCCHK_WriteRegisters2(0, 65, volumes, 2); // CDC_REG_DAC_L_VOLUME_CTRL, CDC_REG_DAC_R_VOLUME_CTRL
CDCCHK_ReadRegisters2(0, 116, &tmp, 1); // CDC_REG_VOL_MICDET_PIN_SAR_ADC
tmp &= ~0x80; // Control volume via register. Do "tmp |= 0x80" to set volume control back to slider.
CDCCHK_WriteRegisters2(0, 116, &tmp, 1);
cdcChkExit();
```

**Comment by gannon345** [comment]
Do we have any idea if/when this might be implemented? The 2ds xl in particular was built with a really flimsy volume adjustment. It's very easy to break when doing a battery replacement (ask me how I know 😅) so this would probably be a useful addition.

**Comment by Anonymous941** [comment]
I've tried to implement this a while ago but I don't know a lot about C so it was very difficult for me.  I can try again, but it should be easy to implement for someone who knows more about it...

**Comment by PabloMK7** [comment]
Currently implementing @profi200 solution.

**Comment by Anonymous941** [comment]
Fixed in #1973
